src/envoy/http/metadata_exchange/config.cc

```diff
@@ -0,0 +1,118 @@
+/* Copyright 2019 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "src/envoy/http/metadata_exchange/config.h"
+#include "common/common/base64.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Wasm {
+namespace MetadataExchange {
+
+// imports from the low-level API
+using Common::Wasm::Null::NullVmPluginFactory;
+using Common::Wasm::Null::Plugin::getMetadataValue;
+using Common::Wasm::Null::Plugin::getRequestHeader;
+using Common::Wasm::Null::Plugin::getResponseHeader;
+using Common::Wasm::Null::Plugin::logDebug;
+using Common::Wasm::Null::Plugin::logInfo;
+using Common::Wasm::Null::Plugin::proxy_setMetadataStruct;
+using Common::Wasm::Null::Plugin::removeRequestHeader;
+using Common::Wasm::Null::Plugin::removeResponseHeader;
+using Common::Wasm::Null::Plugin::replaceRequestHeader;
+using Common::Wasm::Null::Plugin::replaceResponseHeader;
+
+void PluginContext::onCreate() {
+  // TODO(kuat): serialize metadata in the root context instead
+  // populate and encode node metadata
+  auto metadata =
+      getMetadataValue(Common::Wasm::MetadataType::Node, NodeMetadataKey);
+  if (metadata.kind_case() == google::protobuf::Value::kStructValue) {
+    std::string metadata_bytes;
+    metadata.struct_value().SerializeToString(&metadata_bytes);
+    metadata_value_ =
+        Base64::encode(metadata_bytes.data(), metadata_bytes.size());
+  }
+
+  logDebug("metadata_value_ id:" + std::to_string(id()) +
```

Comment:
use fmt::format() in `source/common/common/fmt.h`?

Reply:
using absl::StrCat instead